backend/owa.ts

```diff
@@ -25,31 +32,53 @@ export async function fetchSessionData(partition: string, url: string, interacti
           return;
         }
         finished = true;
-        session.cookies.removeListener('changed', checkCanary);
+        session.cookies.removeListener('changed', onCookie);
         resolve(data);
         if (!popup.isDestroyed()) {
           popup.destroy();
         }
       };
       let checkLoginFinished = async function() {
         try {
+          url = urlObj.href;
           response = await session.fetch(url + 'sessiondata.ashx', { method: 'POST' });
-          finish(await response.json());
+          let json = await response.json();
+          if (isHotmail) {
+            json.owaURL = url;
+          }
+          finish(json);
         } catch (ex) {
         }
       }
-      let checkCanary = function(_event, cookie, _cause, removed) {
-        if (!removed && cookie.name == 'X-OWA-CANARY') {
-          checkLoginFinished();
+      let onCookie = async function(_event, cookie, _cause, removed) {
+        if (removed) {
+          return;
+        }
+        // For Hotmail, check the path to the CANARY cookie.
+        if (cookie.domain == kHotmailServer &&
+          cookie.name == kCookieName &&
+          cookie.path.startsWith("/owa/")) {
+          // Hotmail also sets cookies for /owa/0/, /mail/0/, /calendar/0/ etc.,
+          // but we can use only the /owa/0/ cookie.
+          // We also need to use that URL for the service request.
+          // This needs to happen before CheckLoginFinished().
+          urlObj.hostname = cookie.domain;
+          urlObj.pathname = cookie.path;
+          isHotmail = true;
+        }
+        // If we receive the canary cookie, check whether we're logged in
+        if (cookie.name == kCookieName &&
+            cookie.domain == urlObj.hostname) {
+            checkLoginFinished();
```

Comment:
Missing `await`! This was already wrong in the previous version.

@jermy-c @NeilRashbrook Please pay attention that *every* call to an async function *always* has an `await` in front of it. Or, if you specifically do not want to wait, at the very least a `.catch()`. A call to an async function without these will throw an "unhandled exception", which might kill the entire app, i.e. that's a crash bug.

Reply:
Sorry about that. Thanks for the catch! I'll pay more attention to those.